File: 98. Validate Binary Search Tree.cpp
// 98. Validate Binary Search Tree
// Time: O(N)
// Space: O(N)

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

// preorder traversal
class Solution {
public:
    bool isValidBST(TreeNode* root) {
        bool left = true;
        if (root != NULL) {
            return helper(root->left, LONG_MIN, root->val, left) && helper(root->right, root->val, LONG_MAX, !left);
        } else {
            return true;
        }
    }
    
    bool helper(TreeNode* root, long min, long max, bool left) {
        if (root != NULL) {
            if (left) {
                if (root->val < max && root->val > min) {
                    cout << root->val << endl;
                    return helper(root->left, min, root->val, left) && helper(root->right, root->val, max, !left);
                } else {
                    return false;
                }
            } else {
                if (root->val < max && root->val > min) {
                    cout << root->val << endl;
                    return helper(root->left, min, root->val, !left) && helper(root->right, root->val, max, left);
                } else {
                    return false;
                }
            }
        } else {
            return true;
        }
    }
};

File: 98. Validate Binary Search Tree.py
# 98. Validate Binary Search Tree
# Time: O(N)
# Space: O(logN)

# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

# DFS recursive validate BST
class Solution(object):
    def isValidBST(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        if not root:
            return True
        
        return self.helper(root, float('-inf'), float('inf'))
        
    def helper(self, root, min_val, max_val): # use helper to keep track of min and max for the root.val
        if not root:
            return True
        
        if min_val < root.val < max_val:
            return self.helper(root.left, min_val, root.val) and self.helper(root.right, root.val, max_val)
        
        return False
